Objects/stringobject.c

```diff
@@ -1029,6 +1029,15 @@ string_concat(register PyStringObject *a, register PyObject *bb)
 {
     register Py_ssize_t size;
     register PyStringObject *op;
+    if (a->ob_type != bb->ob_type) {
+        if (Py_Py3kWarningFlag &&
+            PyErr_WarnEx(PyExc_Py3xWarning,
+                    "string concatenation on mixed types is not supported in 3.x:"
+                    "encode or decode the strings to match types", 1) < 0) {
```

Comment:
This is a dynamic check (I think!) so it would be *really* cool IMHO to say "string A is <unicode|string> and string B is <unicode|string>" so the user can then decide which one is the "problem" (so to speak).